opendevin/action/base.py

```diff
@@ -0,0 +1,16 @@
+from dataclasses import dataclass
+
+from ..controller import AgentController
+
+@dataclass
+class Action:
```

Comment:
Do you see this class replacing `Event`? Or living alongside it?

I've been using `Event` to track:
* Actions
* Action Outputs (logs, file contents, webpages)
* User messages
* Thinking in between actions

It's important for the agent to see a stream of these Events (so it can react appropriately), and for the controller to emit a stream of Events (e.g. to show them in the browser).

Reply:
I am actually thinking of replacing `Event` with `Action`:
-  If we want to stream thinking between Actions to the user, you could emit NoopAction (pure text message for thinking).
-  If we want to stream information to the Agent, we can do so by updating the `State`, and feed into the agent? If an action is still running, the agent would be blocked in current implementation so that we would not be able to interrupt it with Event as well?

I think we have some trade-off on the decision here: The advantage is that we don't need to really keep `Action` and `Event` as two seemingly similar but different things (which may confuse people?) - the downside would be the control loop might need to run more frequently, but I think it will get the job down?

Reply:
>  the downside would be the control loop might need to run more frequently, but I think it will get the job down?

Ah, so are you thinking the loop would be more like:
* step 1: `run ls`
* step 2: output of ls
* step 3: read file foo.txt
* step 4: foo contents

instead of the current loop, which combines actions and their outputs into a single step:
* step 1 : `run ls`, output of ls
* step 2: read file foo.txt, foo contents

Reply:
Yes, it roughly look like this:

```python
for i in range(max_iter):
    state = ...
    # iter 2: state has the output of ls
    action = agent.step(state)
    # iter 1: action == `run ls`
    # iter 2: action == "read file foo.txt"

    observation = action.run(self) # self == controller
    # iter 1: observation == output of ls
    # iter 2: observation == "content of foo"

    state = state.update(observation)
```

Reply:
it also allows us to interrupt the control loop easier (since they runs on higher frequency). Also this frequency is also corresponds to LLM's API calling frequency (one step corresponds to 1 LLM call) which could potentially be easier for stats?

Reply:
👍 I like this